In [1]:
import pandas as pd
import numpy as np

In [8]:
census_df = pd.read_csv("./EDA_census_formatted.csv")
census_df.head(10)

Table Name  State Code  Distt. Code Area Name  Total Age-group  \
0      C2308           0            0     INDIA  Total  All ages   
1      C2308           0            0     INDIA  Total       0-6   
2      C2308           0            0     INDIA  Total         7   
3      C2308           0            0     INDIA  Total         8   
4      C2308           0            0     INDIA  Total         9   
5      C2308           0            0     INDIA  Total        10   
6      C2308           0            0     INDIA  Total        11   
7      C2308           0            0     INDIA  Total        12   
8      C2308           0            0     INDIA  Total        13   
9      C2308           0            0     INDIA  Total        14   

   Total Person  Total Males  Total Females  Illiterate Persons  \
0    1210854977    623270258      587584719           447216165   
1     164515253     85752254       78762999           164515253   
2      24826640     12903364       11923276             6748214   
3      26968373     14061937       12906436             4131414   
4      23424638     12214985       11209653             2491904   
5      30552107     16089436       14462671             3454505   
6      24740946     12962604       11778342             1988473   
7      27877307     14637892       13239415             2391529   
8      24280683     12563775       11716908             1810476   
9      25258169     13165128       12093041             2100785   

   Illiterate Males  Illiterate Females  Literate Persons  Literate Males  \
0         188506636           258709529         763638812       434763622   
1          85752254            78762999                 0               0   
2           3388683             3359531          18078426         9514681   
3           2013742             2117672          22836959        12048195   
4           1198339             1293565          20932734        11016646   
5           1661124             1793381          27097602        14428312   
6            929328             1059145          22752473        12033276   
7           1087320             1304209          25485778        13550572   
8            790923             1019553          22470207        11772852   
9            920460             1180325          23157384        12244668   

   Literate Females  
0         328875190  
1                 0  
2           8563745  
3          10788764  
4           9916088  
5          12669290  
6          10719197  
7          11935206  
8          10697355  
9          10912716

## Segmented Analysis

### QUES 1: What percentage of females in the age group 20-24 are illiterate in India, i.e. out of all the females in the age group 20-24, what fraction is illiterate?

In [14]:
# watch_tv_pivot_table = pd.pivot_table(nas_df, columns = ["Watch.TV"], values="Science..", aggfunc=np.mean)

census_illiterate_females_pivot_table = pd.pivot_table(census_df, columns=["Age-group"], values="Illiterate Females", aggfunc=np.sum)
census_illiterate_females_pivot_table

Age-group                 0-6       10       11       12       13       14  \
Illiterate Females  315051996  7173524  4236580  5216836  4078212  4721300   

Age-group                15       16       17       18  ...     60-64  \
Illiterate Females  5693324  5606436  4633656  8783148  ...  51858856   

Age-group              65-69         7     70-74     75-79        8       80+  \
Illiterate Females  38737896  13438124  28795488  13882028  8470688  17735820   

Age-group                 9  Age not stated    All ages  
Illiterate Females  5174260         3700380  1034838116  

[1 rows x 29 columns]

In [17]:
census_df["Total Females"]
# ["Age-group" == "20-24"]

0       587584719
1        78762999
2        11923276
3        12906436
4        11209653
          ...    
3127          804
3128          502
3129          302
3130          329
3131           94
Name: Total Females, Length: 3132, dtype: int64

In [24]:
census_df[(census_df["Age-group"] == "20-24") & (census_df["Total"] == "Total") & (census_df["Area Name"] == "INDIA")]

Table Name  State Code  Distt. Code Area Name  Total Age-group  \
15      C2308           0            0     INDIA  Total     20-24   

    Total Person  Total Males  Total Females  Illiterate Persons  \
15     111424222     57584693       53839529            18679492   

    Illiterate Males  Illiterate Females  Literate Persons  Literate Males  \
15           6467026            12212466          92744730        51117667   

    Literate Females  
15          41627063

In [26]:
perc = 12212466/53839529
perc

0.22683084764727418

### ANS:  22.7 %

### QUES 2: Compare the literacy rates (i.e. the number of literates / total number of population) in each age group and choose the correct option.

In [66]:
census_df.head(100)

Table Name  State Code  Distt. Code                Area Name  Total  \
0       C2308           0            0                    INDIA  Total   
1       C2308           0            0                    INDIA  Total   
2       C2308           0            0                    INDIA  Total   
3       C2308           0            0                    INDIA  Total   
4       C2308           0            0                    INDIA  Total   
..        ...         ...          ...                      ...    ...   
95      C2308           1            0  State - JAMMU & KASHMIR  Total   
96      C2308           1            0  State - JAMMU & KASHMIR  Total   
97      C2308           1            0  State - JAMMU & KASHMIR  Total   
98      C2308           1            0  State - JAMMU & KASHMIR  Total   
99      C2308           1            0  State - JAMMU & KASHMIR  Total   

   Age-group  Total Person  Total Males  Total Females  Illiterate Persons  \
0   All ages    1210854977    623270258      587584719           447216165   
1        0-6     164515253     85752254       78762999           164515253   
2          7      24826640     12903364       11923276             6748214   
3          8      26968373     14061937       12906436             4131414   
4          9      23424638     12214985       11209653             2491904   
..       ...           ...          ...            ...                 ...   
95        13        261079       136728         124351               25759   
96        14        266644       139728         126916               26780   
97        15        270992       140502         130490               32536   
98        16        263041       136077         126964               32564   
99        17        221619       115421         106198               26748   

    Illiterate Males  Illiterate Females  Literate Persons  Literate Males  \
0          188506636           258709529         763638812       434763622   
1           85752254            78762999                 0               0   
2            3388683             3359531          18078426         9514681   
3            2013742             2117672          22836959        12048195   
4            1198339             1293565          20932734        11016646   
..               ...                 ...               ...             ...   
95              9823               15936            235320          126905   
96              9706               17074            239864          130022   
97             11447               21089            238456          129055   
98             10834               21730            230477          125243   
99              8943               17805            194871          106478   

    Literate Females  
0          328875190  
1                  0  
2            8563745  
3           10788764  
4            9916088  
..               ...  
95            108415  
96            109842  
97            109401  
98            105234  
99             88393  

[100 rows x 15 columns]

In [61]:
# watch_tv_pivot_table = pd.pivot_table(nas_df, columns = ["Watch.TV"], values="Science..", aggfunc=np.mean)
# {"Illiterate Persons"/ "Total Person"}

def divideNum(a, b):
    if b == 0:
        return 0
    else:
        return a/b

#     aggfunc= divideNum("Total Person", "Illiterate Persons")
# index=["Area Name"],

census_literate_rate_all_age_groups = pd.pivot_table(census_df, columns=["Age-group"], values=["Illiterate Persons", "Total Person"], aggfunc= np.subtract)
census_literate_rate_all_age_groups

ValueError: invalid number of arguments

### QUES 3: Which state shows the highest female literacy rate?

In [57]:
census_df.head(10)

Table Name  State Code  Distt. Code Area Name  Total Age-group  \
0      C2308           0            0     INDIA  Total  All ages   
1      C2308           0            0     INDIA  Total       0-6   
2      C2308           0            0     INDIA  Total         7   
3      C2308           0            0     INDIA  Total         8   
4      C2308           0            0     INDIA  Total         9   
5      C2308           0            0     INDIA  Total        10   
6      C2308           0            0     INDIA  Total        11   
7      C2308           0            0     INDIA  Total        12   
8      C2308           0            0     INDIA  Total        13   
9      C2308           0            0     INDIA  Total        14   

   Total Person  Total Males  Total Females  Illiterate Persons  \
0    1210854977    623270258      587584719           447216165   
1     164515253     85752254       78762999           164515253   
2      24826640     12903364       11923276             6748214   
3      26968373     14061937       12906436             4131414   
4      23424638     12214985       11209653             2491904   
5      30552107     16089436       14462671             3454505   
6      24740946     12962604       11778342             1988473   
7      27877307     14637892       13239415             2391529   
8      24280683     12563775       11716908             1810476   
9      25258169     13165128       12093041             2100785   

   Illiterate Males  Illiterate Females  Literate Persons  Literate Males  \
0         188506636           258709529         763638812       434763622   
1          85752254            78762999                 0               0   
2           3388683             3359531          18078426         9514681   
3           2013742             2117672          22836959        12048195   
4           1198339             1293565          20932734        11016646   
5           1661124             1793381          27097602        14428312   
6            929328             1059145          22752473        12033276   
7           1087320             1304209          25485778        13550572   
8            790923             1019553          22470207        11772852   
9            920460             1180325          23157384        12244668   

   Literate Females  
0         328875190  
1                 0  
2           8563745  
3          10788764  
4           9916088  
5          12669290  
6          10719197  
7          11935206  
8          10697355  
9          10912716

In [77]:
census_states_literacy_rate = pd.pivot_table(census_df, index=["State Code", "Area Name"], values=["Literate Females", "Total Females"])
census_states_literacy_rate

Literate Females  Total Females
State Code Area Name                                                         
0          INDIA                                  1.512070e+07   2.701539e+07
1          State - JAMMU & KASHMIR                1.288534e+05   2.712938e+05
2          State - HIMACHAL PRADESH               1.051561e+05   1.555278e+05
3          State - PUNJAB                         3.802796e+05   6.024769e+05
4          State - CHANDIGARH                     1.563641e+04   2.182929e+04
5          State - UTTARAKHAND                    1.387239e+05   2.275181e+05
6          State - HARYANA                        3.128699e+05   5.451369e+05
7          State - NCT OF DELHI                   2.548465e+05   3.586490e+05
8          State - RAJASTHAN                      6.706607e+05   1.517124e+06
9          State - UTTAR PRADESH                  2.122418e+06   4.383073e+06
10         State - BIHAR                          9.607600e+05   2.290634e+06
11         State - SIKKIM                         8.904966e+03   1.321871e+04
12         State - ARUNACHAL PRADESH              1.499480e+04   3.079609e+04
13         State - NAGALAND                       2.843572e+04   4.385531e+04
14         State - MANIPUR                        3.993646e+04   6.515899e+04
15         State - MIZORAM                        1.883430e+04   2.491343e+04
16         State - TRIPURA                        5.992708e+04   8.273752e+04
17         State - MEGHALAYA                      4.005177e+04   6.781871e+04
18         State - ASSAM                          3.958316e+05   7.018912e+05
19         State - WEST BENGAL                    1.274458e+06   2.044464e+06
20         State - JHARKHAND                      3.423241e+05   7.382905e+05
21         State - ODISHA                         5.357662e+05   9.545785e+05
22         State - CHHATTISGARH                   3.021623e+05   5.844737e+05
23         State - MADHYA PRADESH                 8.127283e+05   1.609862e+06
24         State - GUJARAT                        8.100453e+05   1.330962e+06
25         State - DAMAN & DIU                    2.931632e+03   4.273379e+03
26         State - DADRA & NAGAR HAVELI           3.710759e+03   6.894207e+03
27         State - MAHARASHTRA                    1.668814e+06   2.488794e+06
28         State - ANDHRA PRADESH                 1.025541e+06   1.937408e+06
29         State - KARNATAKA                      8.339702e+05   1.385225e+06
30         State - GOA                            2.527191e+04   3.307609e+04
31         State - LAKSHADWEEP                    1.127816e+03   1.441379e+03
32         State - KERALA                         6.634906e+05   7.990183e+05
33         State - TAMIL NADU                     1.094116e+06   1.655589e+06
34         State - PUDUCHERRY                     2.114625e+04   2.921572e+04
35         State - ANDAMAN & NICOBAR ISLANDS      5.972598e+03   8.170575e+03

In [70]:
census_df[census_df["State Code"]  == 31]

Table Name  State Code  Distt. Code            Area Name  Total  \
2697      C2308          31            0  State - LAKSHADWEEP  Total   
2698      C2308          31            0  State - LAKSHADWEEP  Total   
2699      C2308          31            0  State - LAKSHADWEEP  Total   
2700      C2308          31            0  State - LAKSHADWEEP  Total   
2701      C2308          31            0  State - LAKSHADWEEP  Total   
...         ...         ...          ...                  ...    ...   
2779      C2308          31            0  State - LAKSHADWEEP  Urban   
2780      C2308          31            0  State - LAKSHADWEEP  Urban   
2781      C2308          31            0  State - LAKSHADWEEP  Urban   
2782      C2308          31            0  State - LAKSHADWEEP  Urban   
2783      C2308          31            0  State - LAKSHADWEEP  Urban   

           Age-group  Total Person  Total Males  Total Females  \
2697        All ages         64473        33123          31350   
2698             0-6          7255         3797           3458   
2699               7          1111          566            545   
2700               8           905          450            455   
2701               9          1188          595            593   
...              ...           ...          ...            ...   
2779           65-69          1094          578            516   
2780           70-74           679          335            344   
2781           75-79           398          191            207   
2782             80+           391          166            225   
2783  Age not stated            88           54             34   

      Illiterate Persons  Illiterate Males  Illiterate Females  \
2697               11920              5100                6820   
2698                7255              3797                3458   
2699                 335               168                 167   
2700                  79                39                  40   
2701                  76                35                  41   
...                  ...               ...                 ...   
2779                 394                91                 303   
2780                 311                79                 232   
2781                 184                55                 129   
2782                 204                46                 158   
2783                  28                16                  12   

      Literate Persons  Literate Males  Literate Females  
2697             52553           28023             24530  
2698                 0               0                 0  
2699               776             398               378  
2700               826             411               415  
2701              1112             560               552  
...                ...             ...               ...  
2779               700             487               213  
2780               368             256               112  
2781               214             136                78  
2782               187             120                67  
2783                60              38                22  

[87 rows x 15 columns]

In [71]:
census_df[census_df["State Code"]  == 2]

Table Name  State Code  Distt. Code                 Area Name  Total  \
174      C2308           2            0  State - HIMACHAL PRADESH  Total   
175      C2308           2            0  State - HIMACHAL PRADESH  Total   
176      C2308           2            0  State - HIMACHAL PRADESH  Total   
177      C2308           2            0  State - HIMACHAL PRADESH  Total   
178      C2308           2            0  State - HIMACHAL PRADESH  Total   
..         ...         ...          ...                       ...    ...   
256      C2308           2            0  State - HIMACHAL PRADESH  Urban   
257      C2308           2            0  State - HIMACHAL PRADESH  Urban   
258      C2308           2            0  State - HIMACHAL PRADESH  Urban   
259      C2308           2            0  State - HIMACHAL PRADESH  Urban   
260      C2308           2            0  State - HIMACHAL PRADESH  Urban   

          Age-group  Total Person  Total Males  Total Females  \
174        All ages       6864602      3481873        3382729   
175             0-6        777898       407459         370439   
176               7        122039        63842          58197   
177               8        117124        61593          55531   
178               9        119100        63574          55526   
..              ...           ...          ...            ...   
256           65-69         12511         6437           6074   
257           70-74          9655         4869           4786   
258           75-79          5748         2879           2869   
259             80+          6694         3126           3568   
260  Age not stated           777          454            323   

     Illiterate Persons  Illiterate Males  Illiterate Females  \
174             1824866            729283             1095583   
175              777898            407459              370439   
176               19485             10007                9478   
177                6140              2983                3157   
178                4563              2289                2274   
..                  ...               ...                 ...   
256                2803               777                2026   
257                2879               811                2068   
258                1891               525                1366   
259                2803               825                1978   
260                 119                63                  56   

     Literate Persons  Literate Males  Literate Females  
174           5039736         2752590           2287146  
175                 0               0                 0  
176            102554           53835             48719  
177            110984           58610             52374  
178            114537           61285             53252  
..                ...             ...               ...  
256              9708            5660              4048  
257              6776            4058              2718  
258              3857            2354              1503  
259              3891            2301              1590  
260               658             391               267  

[87 rows x 15 columns]

In [72]:
census_df[census_df["State Code"]  == 4]

Table Name  State Code  Distt. Code           Area Name  Total  \
348      C2308           4            0  State - CHANDIGARH  Total   
349      C2308           4            0  State - CHANDIGARH  Total   
350      C2308           4            0  State - CHANDIGARH  Total   
351      C2308           4            0  State - CHANDIGARH  Total   
352      C2308           4            0  State - CHANDIGARH  Total   
..         ...         ...          ...                 ...    ...   
430      C2308           4            0  State - CHANDIGARH  Urban   
431      C2308           4            0  State - CHANDIGARH  Urban   
432      C2308           4            0  State - CHANDIGARH  Urban   
433      C2308           4            0  State - CHANDIGARH  Urban   
434      C2308           4            0  State - CHANDIGARH  Urban   

          Age-group  Total Person  Total Males  Total Females  \
348        All ages       1055450       580663         474787   
349             0-6        119434        63536          55898   
350               7         18624         9987           8637   
351               8         17271         9413           7858   
352               9         17719         9666           8053   
..              ...           ...          ...            ...   
430           65-69         15584         8110           7474   
431           70-74         11204         5902           5302   
432           75-79          6347         3212           3135   
433             80+          7525         3745           3780   
434  Age not stated           340          197            143   

     Illiterate Persons  Illiterate Males  Illiterate Females  \
348              250012            115317              134695   
349              119434             63536               55898   
350                3172              1649                1523   
351                1395               738                 657   
352                1397               692                 705   
..                  ...               ...                 ...   
430                3677              1191                2486   
431                2841               954                1887   
432                1426               380                1046   
433                2320               642                1678   
434                  85                35                  50   

     Literate Persons  Literate Males  Literate Females  
348            805438          465346            340092  
349                 0               0                 0  
350             15452            8338              7114  
351             15876            8675              7201  
352             16322            8974              7348  
..                ...             ...               ...  
430             11907            6919              4988  
431              8363            4948              3415  
432              4921            2832              2089  
433              5205            3103              2102  
434               255             162                93  

[87 rows x 15 columns]